### PR TITLE
test: Add test cases to pkg/quay

### PR DESCRIPTION
**Pull request by tkdchen**
* Rewrite part of the test functions in order to add test cases conveniently.
* handleRobotName is refactored slightly. 1) Once the input robotName is matched, no need the check the number of parts separated by +. 2) The regexp is hardcoded and not an input from somewhere else, and the tests ensure that the regexp is correct and can be compiled by the library. So, no need to handle the err of an invalid regexp.
* Unify the type of response data passed to gock JSON func so that string value can be passed as well.
* Assertions are written with gotest.tools assert functions. One of the benefits is the error can be easily checked if it is nil and if a specific err is returned by specifying part of the error message.

### CODE REVIEW

File: pkg/quay/quay.go [hunk 1]
```diff
@@ -120,7 +120,8 @@ func (c *QuayClient) DoesRepositoryExist(organization, imageRepository string) (
 
 	if res.StatusCode == 404 {
 		return false, fmt.Errorf("repository %s does not exist in %s organization", imageRepository, organization)
-	} else if res.StatusCode == 200 {
+	}
+	if res.StatusCode == 200 {
```
Comment (mkosiarc) [inline pkg/quay/quay.go:124]: why this change?

File: pkg/quay/quay.go [hunk 2]
```diff
@@ -509,20 +510,13 @@ func (c *QuayClient) GetAllRobotAccounts(organization string) ([]RobotAccount, e
 // e.g. `org+robot` will be changed to `robot`, `robot` will stay `robot`
 func handleRobotName(robotName string) (string, error) {
 	// Regexp from quay api `^([a-z0-9]+(?:[._-][a-z0-9]+)*)$` with one plus sign in the middle allowed (representing longname)
-	r, err := regexp.Compile(`^[a-z0-9]+(?:[._-][a-z0-9]+)*(?:\+[a-z0-9]+(?:[._-][a-z0-9]+)*)?$`)
+	r, _ := regexp.Compile(`^[a-z0-9]+(?:[._-][a-z0-9]+)*(?:\+[a-z0-9]+(?:[._-][a-z0-9]+)*)?$`)
 	robotName = strings.TrimSpace(robotName)
-	if err != nil {
```
Comment (mkosiarc) [inline pkg/quay/quay.go:514]: we now it compiles now. Is there not a chance that we would update this in the future?

File: pkg/quay/quay.go (hunk 1 above)
Comment (tkdchen) [reply pkg/quay/quay.go]: Restored.

**Review by mkosiarc — APPROVED** [review]

File: pkg/quay/quay_test.go [hunk 3]
```diff
@@ -20,199 +20,461 @@ import (
...
+
 	"github.com/h2non/gock"
 )
 
 const (
 	org       = "test_org"
 	repo      = "test_repo"
 	robotName = "robot_name"
+
+	testRepoName        = "test-repo-using-api"
```
Comment (tnevrlka) [inline pkg/quay/quay_test.go:35]: Is there a need for `testRepoName` instead of using `repo`?

File: pkg/quay/quay_test.go [hunk 4]
```diff
@@ -20,199 +20,461 @@ import (
...
+
+			receivedRepos, err := quayClient.GetAllRepositories("test_org")
+
+			if tc.expectedErr == "" {
+				if err != nil {
+					t.Errorf("Error getting all repositories, Expected nil, got %v", err)
+				}
+				if len(receivedRepos) != 3 {
+					t.Errorf("Possible pagination error, expected 3 repos, got %d repos", len(receivedRepos))
+				}
```
Comment (tnevrlka) [inline pkg/quay/quay_test.go:413]: `assert.NilError` and `assert.Condition` could be used here, couldn't they?

File: pkg/quay/quay_test.go [hunk 5]
```diff
@@ -337,75 +626,92 @@ func TestQuayClient_GetTagsFromPage(t *testing.T) {
...
+				tags, hasAdditional, err := quayClient.GetTagsFromPage(org, repo, page)
+				if tc.expectedErr == "" {
+					assert.NilError(t, err)
+
+					if !reflect.DeepEqual(mockTags, tags) {
+						t.Errorf("tags are not the same, expected `%v`, got `%v`", mockTags, tags)
+					}
+					if hasAdditional != tc.hasAdditional[page-1] {
+						t.Errorf("hasAdditional is not the same, expected `%t`, got `%t`", tc.hasAdditional[page-1], hasAdditional)
+					}
```
Comment (tnevrlka) [inline pkg/quay/quay_test.go:654]: `assert.DeepEqual` and `assert.Condition` might work here?

**Review by tnevrlka — APPROVED** [review]
Thank you, very nice

File: pkg/quay/quay_test.go [hunk 6]
```diff
@@ -20,199 +20,461 @@ import (
...
-		MatchHeader("Authorization", "Bearer authtoken").
-		Post("/repository").
-		Reply(200).JSON(map[string]string{
-		"description": "description",
-		"namespace":   "redhat-appstudio-user",
-		"name":        "test-repo-using-api",
-	})
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer gock.Off()
```
Comment (Michkov) [inline pkg/quay/quay_test.go:108]: defer is called at the end of the function, so it can be just called once at beginning. Am I right @mmorhun

File: pkg/quay/quay_test.go (hunk 3 above)
Comment (tkdchen) [reply pkg/quay/quay_test.go]: Good catch. This was refactored from corresponding test. Updated to use `repo`.

File: pkg/quay/quay_test.go (hunk 4 above)
Comment (tkdchen) [reply pkg/quay/quay_test.go]: Updated. PTAL.

File: pkg/quay/quay_test.go (hunk 6 above)
Comment (tkdchen) [reply pkg/quay/quay_test.go:107]: Calling defer in the anonymous function intentionally due to some test function failed without it. But, I forget which one :sweat_smile:

File: pkg/quay/quay.go [hunk 7]
```diff
@@ -559,20 +559,13 @@ func (c *QuayClient) GetAllRobotAccounts(organization string) ([]RobotAccount, e
 // e.g. `org+robot` will be changed to `robot`, `robot` will stay `robot`
 func handleRobotName(robotName string) (string, error) {
 	// Regexp from quay api `^([a-z0-9]+(?:[._-][a-z0-9]+)*)$` with one plus sign in the middle allowed (representing longname)
-	r, err := regexp.Compile(`^[a-z0-9]+(?:[._-][a-z0-9]+)*(?:\+[a-z0-9]+(?:[._-][a-z0-9]+)*)?$`)
+	r, _ := regexp.Compile(`^[a-z0-9]+(?:[._-][a-z0-9]+)*(?:\+[a-z0-9]+(?:[._-][a-z0-9]+)*)?$`)
```
[the remark this answers is not in the document]
Comment (tkdchen) [reply pkg/quay/quay.go]: Updated.

File: pkg/quay/quay_test.go (hunk 5 above)
Comment (tkdchen) [reply pkg/quay/quay_test.go]: Updated.
